### DISCUSSION TITLE
.NET 8.0 headless testing & transition

### DISCUSSION THREAD

**Discussion by Frooxius** [Yellow-Dog-Man/Resonite-Issues#2539] [Pre-release Testing]
Hello everyone!

Guess what... another test!

## What is being tested
We've moving headless from .NET Framework to the modern .NET 8.0!

This should provide much better performance for headless and remove dependency on Mono - instead you need to use the modern .NET 8.0 runtime to run it.

https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2265

## How to test
If you have access to headless builds, you can find the .NET 8.0 variant in a folder named `net8.0`. When you can, try hosting your sessions on this version instead.

The version is compatible with normal builds (and normal headlesses) and will be included with every build - there is no special `prerelease` branch for this one, so you can freely switch back and forth.

If you notice any breakage and issues that are NOT present on the normal .NET Framework headless, report them please!

## If you find something that got broken
- Make sure that the issue doesn't occur on the .NET Framework headless
- Create a GitHub issue with the prerelease: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/new?assignees=shiftyscales&labels=prerelease&projects=&template=prerelease.yml

## If you find that stuff works okay
Please report it too and include details - how many users, where is the host, where were users (if you know), how long did the session run for, what did you (roughly) do and such. This way we know things are working okay.

## How long will testing go for?
At least a week, more if there are major issues. We will provide both versions of headless for the duration and then switch and provide only .NET 8.0 headless

## Phasing out the old headless
Please note that this means the old headless will be phased out and you should start preparing to migrate your hosting and setup to this one instead! 

Right now we're keeping it for at least a week to transition, but if you need more time, let us know in a comment blow and we can extend this (within reasonable limits).

**Comment by ghost** [comment]
Is it windows only atm?

**Reply by Frooxius** [reply]
Should be Winodws and x64 Linuux

**Comment by gameboycjp** [comment]
Would testing on Linux ARM be appreciated?

**Reply by Frooxius** [reply]
We don't support ARmM

**Comment by troyBORG** [comment]
What do we need to do to run it with linux?
Replace `mono` with `dotnet` in our launch script and make sure we have the dotnet8 package installed...

So:
`dotnet ~/Steam/steamapps/common/Resonite/Headless//net8.0/Resonite.exe`

But I'm getting:
```
Error:
  An assembly specified in the application dependencies manifest (Resonite.deps.json) has already been found but with a different file extension:
    package: 'Resonite', version: '1.0.0'
    path: 'Resonite.dll'
    previously found assembly: '/home/steam/Steam/steamapps/common/Resonite/Headless/net8.0/Resonite.exe
```

**Reply by bredo228** [reply]
Right now it'll work if you run `dotnet Resonite.dll`

**Comment by troyBORG** [comment]


**Reply by Readun** [reply]
Well... Mods are mods... They can break and should try to be reported to the people who made them.
With official stuff, always try to report bugs while using and replicating your bugs while running everything vanilla.

**Reply by troyBORG** [reply]
I did I made a report about it on there ModLoader.

But I guess its hard to see.  But when it boots it doesn't sign me in or load the Config file.
[noMods.log](https://github.com/user-attachments/files/16190176/noMods.log)

**Comment by troyBORG** [comment]
[new.log](https://github.com/user-attachments/files/16190360/new.log)

Ok,  I got an error I'm having when trying on my Ubuntu Linux Server.
(It does this error as soon as you try to login.   This error log is from when I had it login via config file)

`System.IO.DirectoryNotFoundException: Could not find a part of the path '/home/steam/Steam/steamapps/common/Resonite/Headless/net8.0/Migrations/NVR.json'`

And there is no Migrations folder in that net8.0 folder.
![image](https://github.com/user-attachments/assets/57fd7ade-7e43-4da7-9eba-3c1cca5a65a1)

It looks like logins might be working now without giving that error.

**Reply by troyBORG** [reply]
Ok.  Well its not doing that anymore.  And it seems like I can login now..  But it gets stuck and like doesn't start opening the world...
[03_20_26.log](https://github.com/user-attachments/files/16190815/03_20_26.log)

**Reply by LexiBasilisk** [reply]
As requested in the instructions above- please report issues via the prerelease issue template, @troyBORG.

https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/new?assignees=shiftyscales&labels=prerelease&projects=&template=prerelease.yml

**Comment by bredo228** [comment]
Ran mentor workshop off .NET 8 headless, had a peak of around 16 people from varying places worldwide doing builder things etc. after 2 hours, session was still running fine 

DeSync lounge is also on .NET 8 now - haven't noticed any issues cropping up yet with people coming in and out and having a bunch of the regular chaos happening over time. 

DeSync lounge is currently sitting at 8 users after 11h of being up (not all this time with people in it), using 30% of a core worth of cpu and 4.76GiB of RAM on Ryzen 9 5950x

**Reply by sveken** [reply]
To add to this Comparing the metrics on desync to other periods of usage it seems to be running a good 20% less CPU load wise too while people are playing

Idle has dropped from 7% to 4% and on idle over 600MB less memory usage

**Reply by Frooxius** [reply]
Ooo thank you for the info!

**Comment by j4lc** [comment]
Hosting the French-speaking meetup, no issues at all to be reported.

**Reply by Frooxius** [reply]
Eee yay! Thanks for report!

**Comment by FoxoBread** [comment]
Happy to report no issues while running two headless instances on Linux yesterday! :3

**Comment by Frooxius** [comment]
## Removing the old .NET Framework / Mono headless
Hey everyone!

Thank you for all the testing so far, we've been getting really good results and only relatively minor issues (which will be fixed up soon). With this, we'd like to remove the old .NET Framework / Mono headless as soon as we can and make full switch to .NET 8.0.

Before we do that, I'd like to ask - how long do you need to prepare for this switch or do more testing?

Right now, my plan would be to leave this open up for a week (meaning we'll remove the old headless next Monday) - unless you need more time - in which case please let us know in a comment bellow this!

**Reply by shadowpanther** [reply]
When the old mono build is removed, would the new 8.0 build be moved in its place?

I'm planning to update my headless image and need to know - should I expect the new executable to remain where it is right now or should I wait for the removal to happen and then update my image with the executable in its final place?

**Reply by Frooxius** [reply]
@shadowpanther Yes, it'll be placed where the Mono build is right now and the net8.0 subfolder will be removed.

**Reply by gentlecolts** [reply]
Hey, from what I've heard there's not a lot to do to upgrade, but I haven't been following this feature at all.  Has a migration guide been posted somewhere?  Does anything need to be done once it's out of beta?  I'd like to update my personal docker image asap.

**Reply by Frooxius** [reply]
@gentlecolts Once it's out of beta you'll need to make sure to update your launch/setup scripts to use .NET 8.0 runtime instead of Mono. Other than that there isn't super much.

**Reply by Frooxius** [reply]
Also just a heads up, I'll give it a few more days since last week has been... crazy and I didn't get to address some of the stuff I wanted to.

However since there aren't really any requests to delay this, I'll make the switch to .NET 8.0 as soon as those things are resolved.

**Reply by shadowpanther** [reply]
> Hey, from what I've heard there's not a lot to do to upgrade, but I haven't been following this feature at all. Has a migration guide been posted somewhere? Does anything need to be done once it's out of beta? I'd like to update my personal docker image asap.

Instead of `mono Resonite.exe`, you do `dotnet Resonite.dll`. That's all. Well, and you need a `dotnet-runtime-8.0` (or whatever your distro calls it) for the `dotnet` command itself.

I've made a test branch of my docker image. If all goes well, I'd only need to change the path to the `Resonite.dll` when the mono release gets removed.
Source: https://github.com/shadowpanther/resonite-headless/tree/dotnet8
Docker hub image: `shadowpanther/resonite-headless:dotnet8`

**Reply by j4lc** [reply]
I also wrote a wiki page earlier today, just in case: https://wiki.resonite.com/Headless_Server_Software/Migration

**Reply by sveken** [reply]
> I also wrote a wiki page earlier today, just in case: https://wiki.resonite.com/Headless_Server_Software/Migration

Also if your image is already using dotnet8 once we migrate over the paths will need to change. But that should be it.

**Reply by gentlecolts** [reply]
Something pretty significant that seems to be missing is that there's a separate config folder in the net8.0 folder that the new headless uses, regardless of launch working directory.  Is the net8.0 folder going away once changes are finished?

**Reply by Frooxius** [reply]
@gentlecolts I'm not sure if I understand about the config folder. It should be the same as with current Mono one.

The net8.0 folder will be gone and the .NET 8.0 headless will be placed where the Mono one is right now.

**Reply by gentlecolts** [reply]
> It should be the same as with current Mono one

I just updated mine and I had to copy my config.json from `Headless/Config` to `Headless/net8.0/Config` to get it working again

**Reply by LexiBasilisk** [reply]
I believe what Frooxius had meant- once it's live- it should be the same `Headless/Config` folder that is already in use. The `.net8.0` folder just exists for the sake of this testing / transitory period, and afterward, the contents of that folder will be in the main `Headless` folder replacing the existing mono version. @gentlecolts

**Reply by shadowpanther** [reply]
My image passes separate `Logs` and `Config` paths to the headless (those are passed into the container as volumes, mounted from your host system), and the start script tries to look both in `net8.0` and in the main folder, so it should work in both cases.

After the `net8.0` is made the main and only release, I'll merge the branch into the main (`:latest`) image, which should also allow everyone on the `mono` headless to (hopefully) seamlessly migrate without changing anything outside the image.

**Comment by Frooxius** [comment]
Just a heads up, I'll be removing the old .NET Framework / Mono headless in an update today and the .NET 8.0 headless will move to where that one was!

**Reply by shadowpanther** [reply]
Okay, I've merged the changes into the `master` branch of my image repo, and the `:latest` image should be ready in a few minutes.

It'd try to launch the `dotnet8` build from both locations, so it should just work after the move.

**Comment by Frooxius** [comment]
This is now closed and .NET 8.0 headless is the main! Thank you everyone for testing!
